src/stream/stream/fold.rs

```diff
@@ -0,0 +1,53 @@
+use std::marker::PhantomData;
+use std::pin::Pin;
+
+use crate::future::Future;
+use crate::stream::Stream;
+use crate::task::{Context, Poll};
+
+#[doc(hidden)]
+#[allow(missing_debug_implementations)]
+pub struct FoldFuture<S, F, T, B> {
+    stream: S,
+    f: F,
+    acc: Option<B>,
+    __t: PhantomData<T>,
+}
+
+impl<S, F, T, B> FoldFuture<S, F, T, B> {
+    pin_utils::unsafe_pinned!(stream: S);
+    pin_utils::unsafe_unpinned!(f: F);
+    pin_utils::unsafe_unpinned!(acc: Option<B>);
+
+    pub(super) fn new(stream: S, init: B, f: F) -> Self {
+        FoldFuture {
+            stream,
+            f,
+            acc: Some(init),
+            __t: PhantomData,
+        }
+    }
+}
+
+impl<S, F, B> Future for FoldFuture<S, F, S::Item, B>
+where
+    S: Stream + Unpin + Sized,
+    F: FnMut(B, S::Item) -> B,
+{
+    type Output = B;
+
+    fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let next = futures_core::ready!(self.as_mut().stream().poll_next(cx));
+
+        match next {
+            Some(v) => {
+                cx.waker().wake_by_ref();
```

Comment:
Instead of waking ourselves on every step, let's turn this into a loop instead and implement `poll()` like here:

https://github.com/rust-lang-nursery/futures-rs/blob/master/futures-util/src/stream/fold.rs